### PR TITLE
Support HuggingFace kernels package for native diffusion attention backends

### PR DESCRIPTION
**Pull request by SamitHuang**
## Purpose

This PR implements native support for the HuggingFace `kernels` library to resolve the train ↔ rollout consistency gap in diffusion-based reinforcement learning (such as GRPO) workflows.

Currently, during RL rollout generation (e.g., text-to-image or text-to-video), vLLM-Omni uses local, static implementations of attention backends. During the training phase, standard training frameworks (like HuggingFace Diffusers) often fetch highly optimized, custom compute kernels dynamically from the Hugging Face Hub (the `kernels` community repository). These tiny numerical and behavioral differences can compound iteratively across multiple denoising steps, leading to divergence in policy updates and reward scoring.

To address this, we integrate the HuggingFace `kernels` loader directly into vLLM-Omni as two new first-class attention backend options:
1. `FLASH_ATTN_HUB`: Loads FlashAttention-2 (`kernels-community/flash-attn2`) dynamically from the Hub.
2. `FLASH_ATTN_3_HUB`: Loads FlashAttention-3 (`kernels-community/flash-attn3`) dynamically from the Hub (optimized for Hopper+ SM90 architectures).

On non-CUDA platforms, or when the `kernels` package is not installed, the platform manager logs a warning and falls back gracefully to local `FLASH_ATTN` implementations.

## Test Plan

1. Added unit tests in `tests/diffusion/attention/test_kernels_hub.py` covering:
Run unit tests locally:
```bash
pytest tests/diffusion/attention/test_kernels_hub.py -s
```
2. E2E test using z-image t2i generation with kernels FA3.

## Test Result

All unit tests pass successfully:
```
tests/diffusion/attention/test_kernels_hub.py INFO 07-06 12:51:51 [scheduler.py:239] Chunked prefill is enabled with max_num_batched_tokens=2048.
INFO 07-06 12:51:51 [vllm.py:977] Asynchronous scheduling is enabled.
INFO 07-06 12:51:51 [kernel.py:270] Final IR op priority after setting platform defaults: IrOpPriorityConfig(rms_norm=['native'], fused_add_rms_norm=['native'])
--- Running test: test_kernels_hub_platform_fallback
WARNING 07-06 12:51:52 [platform.py:130] HuggingFace `kernels` library is not available. Falling back to local FLASH_ATTN.
WARNING 07-06 12:51:52 [platform.py:136] HuggingFace `kernels` library is not available. Falling back to local FLASH_ATTN.
.--- Running test: test_kernels_hub_execution
INFO 07-06 12:51:53 [flash_attn_hub.py:63] Loading flash-attn2 kernel from HuggingFace Hub...
Fetching ... files: 0it [00:00, ?it/s]Fetching ... files: 23it [00:00, 4806.63it/s]
✓ FlashAttentionHub PASSED, max diff: 0.00390625
INFO 07-06 12:51:54 [flash_attn_hub.py:247] Loading flash-attn3 kernel from HuggingFace Hub...
Warning: You are sending unauthenticated requests to the HF Hub. Please set a HF_TOKEN to enable higher rate limits and faster downloads.
Fetching ... files: 0it [00:00, ?it/s]Fetching ... files: 8it [00:00, 13992.67it/s]
✓ FlashAttention3Hub PASSED, max diff: 0.00390625
.
======================== 2 passed, 20 warnings in 5.10s ========================
```
kernels fa3:
image

torch spda:
image

### CODE REVIEW

**Comment by SamitHuang** [comment]
@RuixiangMa  @wtomin PTAL

File: docs/user_guide/diffusion/attention_backends.md [hunk 1]
```diff
@@ -210,6 +212,36 @@ Notes:
...
+The following backend options are supported:
+- `FLASH_ATTN_HUB` (HuggingFace `kernels-community/flash-attn2`)
+- `FLASH_ATTN_3_HUB` (HuggingFace `kernels-community/flash-attn3`)
+
+### Installation
+
+To use these backends, you must install the `kernels` library:
+
+```bash
+pip install kernels
```
Comment (wtomin) [inline docs/user_guide/diffusion/attention_backends.md]: Any versions or minimal versions requirements for `kernels` to work with vllm-omni and verl-omni?

File: tests/diffusion/attention/test_kernels_hub.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    # Test FLASH_ATTN_HUB falls back to FLASH_ATTN
+    backend_path = CudaOmniPlatform.get_diffusion_attn_backend_cls("FLASH_ATTN_HUB", head_size=64)
+    assert backend_path == DiffusionAttentionBackendEnum.FLASH_ATTN.get_path()
+
+    # Test FLASH_ATTN_3_HUB falls back to FLASH_ATTN
+    backend_path = CudaOmniPlatform.get_diffusion_attn_backend_cls("FLASH_ATTN_3_HUB", head_size=64)
+    assert backend_path == DiffusionAttentionBackendEnum.FLASH_ATTN.get_path()
+
+
+@pytest.mark.skipif(not is_cuda, reason="Hub kernels execution requires CUDA platform")
```
Comment (wtomin) [inline tests/diffusion/attention/test_kernels_hub.py]:
```suggestion
@pytest.mark.skipif(not is_cuda, reason="Hub kernels execution requires CUDA platform")
@pytest.mark.core_model
@hardware_test(res={"cuda": "L4"}, num_cards=1)
```

If you want to include this test script in L1 test, these markers are needed.
However, I am not sure if the CI docker image has `kernels` installed, so maybe ignore my comment here.

File: vllm_omni/diffusion/attention/backends/flash_attn_hub.py [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+        self.qkv_layout = qkv_layout
+        if backend_kwargs:
+            logger.warning("FlashAttentionHubImpl ignoring backend_kwargs: %s", list(backend_kwargs.keys()))
+
+        # Lazily get the kernel from the Hub to avoid network/disk overhead on import
+        from kernels import get_kernel
+
+        logger.info("Loading flash-attn2 kernel from HuggingFace Hub...")
+        try:
+            hub_module = get_kernel("kernels-community/flash-attn2", version=1)
```
Comment (wtomin) [inline vllm_omni/diffusion/attention/backends/flash_attn_hub.py:96]: I don't quick understand the differences between `version=1`, `version=2`, and default. Can you explain their differences and the fallback logic in the document?

File: vllm_omni/diffusion/attention/backends/flash_attn_hub.py [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+        value: torch.Tensor,
+        attn_metadata: AttentionMetadata = None,
+    ) -> torch.Tensor:
+        attention_mask = attn_metadata.attn_mask if attn_metadata is not None else None
+        full_attn_spans = attn_metadata.full_attn_spans if attn_metadata is not None else None
+
+        # Try piecewise attention
+        if full_attn_spans is not None:
+            logger.debug("Using piecewise Flash Attention for mixed causal/full mask")
+            attn_func = partial(
```
Comment (RuixiangMa) [inline vllm_omni/diffusion/attention/backends/flash_attn_hub.py:191]: If the hub kernel only provides flash_attn_varlen_func but not flash_attn_func (which the __init__ allows — it only requires one of the two), this would pass None as attn_func and crash when piecewise_attn calls it.

File: vllm_omni/platforms/cuda/platform.py [hunk 5]
```diff
@@ -121,6 +121,22 @@ def get_diffusion_attn_backend_cls(
 
         if selected_backend is not None:
             backend_upper = selected_backend.upper()
+            if backend_upper in ("FLASH_ATTN_HUB", "FLASH_ATTN_3_HUB"):
```
Comment (RuixiangMa) [inline vllm_omni/platforms/cuda/platform.py:124]: The PR description states FA3 Hub is "optimized for Hopper+ SM90 architectures," but  doesn't gate FLASH_ATTN_3_HUB on SM90+？

File: tests/diffusion/attention/test_kernels_hub.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+                num_heads=num_heads, head_size=head_dim, softmax_scale=1.0 / (head_dim**0.5), causal=False
+            )
+            output_fa3_hub = fa3_hub_impl.forward(q.clone(), k.clone(), v.clone(), attn_metadata_sdpa)
+            assert output_fa3_hub.shape == q.shape
+            assert not torch.isnan(output_fa3_hub).any()
+            # Verify correctness against reference
+            max_diff = torch.max(torch.abs(output_ref - output_fa3_hub)).item()
+            assert max_diff < 1e-2, f"FlashAttention3Hub output differs too much from SDPA reference: {max_diff}"
+            print(f"✓ FlashAttention3Hub PASSED, max diff: {max_diff}")
+        except Exception as e:
```
Comment (RuixiangMa) [inline tests/diffusion/attention/test_kernels_hub.py]: the try/except around kernel loading just print()s a message and continues. The test passes even if the hub kernel can't be loaded ( package not installed, etc.). This means CI shows green without actually testing the hub kernels.

File: tests/diffusion/attention/test_kernels_hub.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+from vllm_omni.diffusion.attention.backends.abstract import AttentionMetadata
+from vllm_omni.diffusion.attention.backends.registry import DiffusionAttentionBackendEnum
+from vllm_omni.diffusion.attention.backends.sdpa import SDPAImpl
+from vllm_omni.platforms import current_omni_platform
+
+is_cuda = current_omni_platform.is_cuda()
+
+
+def test_kernels_hub_platform_fallback(monkeypatch: pytest.MonkeyPatch):
```
Comment (yenuo26) [inline tests/diffusion/attention/test_kernels_hub.py:19]: please add mark

File: tests/diffusion/attention/test_kernels_hub.py (hunk 2 above)
Comment (yenuo26) [inline tests/diffusion/attention/test_kernels_hub.py]: In practice, this kind of skipif can easily go unnoticed — if the resources are misconfigured by accident, no one will realize that the test case was skipped and not actually executed. Since you have already added the hardware marker, I would suggest removing this skipif.

File: tests/diffusion/attention/test_kernels_hub.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+    # 1. Test PyTorch SDPA reference
+    sdpa_impl = SDPAImpl(num_heads=num_heads, head_size=head_dim, softmax_scale=1.0 / (head_dim**0.5), causal=False)
+    attn_metadata_sdpa = AttentionMetadata(attn_mask=None)
+    output_ref = sdpa_impl.forward(q.clone(), k.clone(), v.clone(), attn_metadata_sdpa)
+
+    # 2. Test FlashAttentionHubBackend (FlashAttention 2)
+    from vllm_omni.diffusion.attention.backends.flash_attn_hub import FlashAttentionHubImpl
+
+    try:
```
Comment (yenuo26) [inline tests/diffusion/attention/test_kernels_hub.py]: I believe that since the purpose of this test case is specifically to verify this functionality, it should perform strict validation here, rather than using try-except, which could cause a false pass in the CI.

File: vllm_omni/diffusion/attention/backends/flash_attn_hub.py (hunk 4 above)
Comment (SamitHuang) [reply vllm_omni/diffusion/attention/backends/flash_attn_hub.py:191]: thanks, fixed by add a shared _run_varlen_dense() helper

File: vllm_omni/platforms/cuda/platform.py (hunk 5 above)
Comment (SamitHuang) [reply vllm_omni/platforms/cuda/platform.py:124]: fixed. logic: 
  1. kernels missing → FLASH_ATTN (unchanged)
  2. GPU < sm_90 → FLASH_ATTN_HUB (new)
  3. GPU >= sm_90 → FLASH_ATTN_3_HUB

File: vllm_omni/diffusion/attention/backends/flash_attn_hub.py (hunk 3 above)
Comment (SamitHuang) [reply vllm_omni/diffusion/attention/backends/flash_attn_hub.py:96]: version=1 / version=2 refer to major API versions of the HF Hub kernel repo (kernels-community/flash-attn2), stored as git branches v1, v2, etc. unrelated to choosing FLASH_ATTN_HUB (FA2) vs FLASH_ATTN_3_HUB (FA3).

File: tests/diffusion/attention/test_kernels_hub.py (hunk 7 above)
Comment (SamitHuang) [reply tests/diffusion/attention/test_kernels_hub.py:19]: added

**Review by wtomin — APPROVED** [review]
LGTM.

**Comment by wtomin** [comment]
I think npu-ci failed at docker image launching, not relevant to this PR.
